### PR TITLE
added automatic portChange feature if current port is busy

### PR DESCRIPTION
**Pull request by EraKin575**
## Related Issue
  - Info about Issue or bug

Closes: #1051 
#### Describe the changes you've made
added a automatic port changing feature for test and record mode if current port is busy
image

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Code style update (formatting, local variables)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Please let us know if any test cases are added

Please describe the tests(if any). Provide instructions how its affecting the coverage.

#### Describe if there is any unusual behaviour of your code(Write `NA` if there isn't)
A clear and concise description of it.

## Checklist:

- [ ] My code follows the style guidelines of this project.
- [ ] I have performed a self-review of my own code.
- [ ] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have made corresponding changes to the documentation.
- [ ] My changes generate no new warnings.
- [ ] I have added tests that prove my fix is effective or that my feature works.
- [ ] New and existing unit tests pass locally with my changes.

## Screenshots (if any)

|        Original         |          Updated           |
|:-----------------------:|:--------------------------:|
| **original screenshot** | <b>updated screenshot </b> |

### CODE REVIEW

**Review by nakulbh — COMMENTED** [review]
I Things there is no better way than this to change port automatically, Looks good to me

**Comment by AkashKumar7902** [comment]
What if the next port is in use by some other running service? A while loop with a random function to generate port would be better I believe

**Comment by nakulbh** [comment]
> What if the next port is in use by some other running service? A while loop with a random function to generate port would be better I believe

it will loop over until it will find a free port

File: pkg/proxy/proxy.go [hunk 1]
```diff
@@ -348,6 +348,9 @@ func BootProxy(logger *zap.Logger, opt Option, appCmd, appContainer string, pid
 	if opt.Port == 0 {
 		opt.Port = 16789
 	}
+	for !isPortAvailable(opt.Port) {
```
Comment (charankamarapu) [inline pkg/proxy/proxy.go:351]: Define upper and lower bounds. For loop running continuously is not preferred, define max attempts.

**Review by charankamarapu — COMMENTED** [review]
Please address the comments

File: pkg/proxy/proxy.go [hunk 2]
```diff
@@ -349,7 +349,19 @@ func BootProxy(logger *zap.Logger, opt Option, appCmd, appContainer string, pid
...
 	}
+	maxAttempts := 1000
+	attemptsDone:=0
 
+
+    if !isPortAvailable(opt.Port) {
+		for i := 1024; i <= 65535 && attemptsDone < maxAttempts; i++ {
+			if isPortAvailable(uint32(i)) {
+				opt.Port = uint32(i)
+				break
```
Comment (charankamarapu) [inline pkg/proxy/proxy.go:360]: If all ports are busy you should throw an error right ..? missed this in last review my bad.

**Review by charankamarapu — COMMENTED** [review]
Please address the comment

File: pkg/proxy/proxy.go [hunk 3]
```diff
@@ -349,7 +349,26 @@ func BootProxy(logger *zap.Logger, opt Option, appCmd, appContainer string, pid
...
 	}
+	maxAttempts := 1000
+	attemptsDone:=0
 
+
+    if !isPortAvailable(opt.Port) {
+		for i := 1024; i <= 65535 && attemptsDone < maxAttempts; i++ {
+			if isPortAvailable(uint32(i)) {
+				opt.Port = uint32(i)
+				logger.Info(Emoji+"Port %v is already in use, switching to next available port");
```
Comment (charankamarapu) [inline pkg/proxy/proxy.go:360]: I don't think we need emoji here it is already appended when we initialise logger and include the port number which you are switching to in the log.

File: pkg/proxy/proxy.go [hunk 4]
```diff
@@ -349,7 +349,26 @@ func BootProxy(logger *zap.Logger, opt Option, appCmd, appContainer string, pid
...
+
+				break
+			}
+			attemptsDone++
+		}
+	}
+	
+	if maxAttempts<=attemptsDone{
+		logger.Error(Emoji+"Failed to find an available port to start proxy server")
```
Comment (charankamarapu) [inline pkg/proxy/proxy.go:369]: Remove the emoji.

File: pkg/proxy/proxy.go [hunk 5]
```diff
@@ -349,7 +349,26 @@ func BootProxy(logger *zap.Logger, opt Option, appCmd, appContainer string, pid
...
+				opt.Port = uint32(i)
+				logger.Info(Emoji+"Port %v is already in use, switching to next available port");
+
+				break
+			}
+			attemptsDone++
+		}
+	}
+	
+	if maxAttempts<=attemptsDone{
```
Comment (charankamarapu) [inline pkg/proxy/proxy.go:368]: beautify the statement I guess there must be spaces before and after the <= symbol. There must be some extension in the IDE you can use it and make sure it is indentation is correct.

**Review by charankamarapu — COMMENTED** [review]
Please address the comments, rest looks fine.

File: pkg/proxy/proxy.go [hunk 6]
```diff
@@ -349,11 +349,30 @@ func BootProxy(logger *zap.Logger, opt Option, appCmd, appContainer string, pid
...
+	
+	if maxAttempts <= attemptsDone{
+		logger.Error("Failed to find an available port to start proxy server")
+		return nil
+	}
 	//IPv4
 	localIp4, err := util.GetLocalIPv4()
 	if err != nil {
-		log.Fatalln(Emoji+"Failed to get the local Ip4 address", err)
+		log.Fatalln("Failed to get the local Ip4 address", err)
```
Comment (charankamarapu) [inline pkg/proxy/proxy.go:375]: Why is it removed from here? This is not zap.logger, It is just log. Here we should use emoji.

**Review by charankamarapu — COMMENTED** [review]
Please address the comment.

File: pkg/proxy/proxy.go [hunk 7]
```diff
@@ -349,7 +348,26 @@ func BootProxy(logger *zap.Logger, opt Option, appCmd, appContainer string, pid
 	if opt.Port == 0 {
 		opt.Port = 16789
 	}
+	maxAttempts := 1000
+	attemptsDone:=0
+
 
+    if !isPortAvailable(opt.Port) {
+		logger.Info("Port is already in use, switching to next available port",zap.Uint32("port",opt.Port));
```
Comment (charankamarapu) [inline pkg/proxy/proxy.go:356]: `Port is already in use, switching to next available port {port : 16789}` this is how the logger looks when you run keploy. It gives a sense that you you found that port is already in use and you switching to port 16789 which is wrong. You can delete this log or correct it . Deleting this log makes more sense for me.

**Review by charankamarapu — APPROVED** [review]
ship it!

**Pull request merged**